lib/mutgen.c

```diff
@@ -25,66 +25,350 @@
+     * https://gcc.gnu.org/bugzilla/show_bug.cgi?id=81631 */
+    const mutation_t *ia = *(const mutation_t **) (uintptr_t) a;
+    const mutation_t *ib = *(const mutation_t **) (uintptr_t) b;
+    out = ia->id < 0 || ib->id < 0 ? 0 : (ia->id - ib->id) - (ib->id - ia->id);
+    if (out == 0)
+        out = (ib->time > ia->time) - (ib->time < ia->time);
+    if (out == 0)
+        out = (ia->parent > ib->parent) - (ia->parent < ib->parent);
+    return out;
+}
 
-static const mutation_type_t binary_mutation_types[] = {
-    {"0", "1"}
+static void 
+insert_mutation(site_t *site, mutation_t *new)
+{
+    mutation_t *u;
+    u = site->mutations;
+    new->prev = NULL;
+    new->next = u;
+    if (u != NULL) {
+        u->prev = new;
+    }
+    site->mutations = new;
+    site->mutations_length++;
+    assert(site->mutations_length > 0);
+}
+
+static void 
+delete_mutation(site_t *site, mutation_t *mut)
+{
+    mutation_t *u;
+    // TODO: need to dealloc mut?
+    u = mut->prev;
+    if (u == NULL) {
+        site->mutations = mut->next;
+    } else {
+        u->next = mut->next;
+    }
+    site->mutations_length--;
+}
+
+static void
+sort_mutations(site_t *site)
+{
+    size_t k;
+    mutation_t *m;
+    mutation_t *p[site->mutations_length];
+    size_t num_mutations = site->mutations_length;
+    k = 0;
+    for (m = site->mutations; m != NULL; m = m->next) {
+        p[k] = m;
+        k++;
+    }
+    assert(k == num_mutations);
+    qsort(p, (size_t) num_mutations, sizeof(mutation_t *), &cmp_mutationp);
+    site->mutations = p[0];
+    for (k = 0; k < num_mutations; k++) {
+        m = p[k];
+        if (k == 0) {
+            m->prev = NULL;
+        } else {
+            m->prev = p[k-1];
+        }
+        if (k == num_mutations - 1) {
+            m->next = NULL;
+        } else {
+            m->next = p[k+1];
+        }
+    }
+}
+
+/*******************
+ * Mutation model */
+
+/* We will use these as dummy values, that should all be replaced. */
+static const char dummy_alphabet[2] = "XY";
+
+// TODO: why do alphabets have const but other things not?
+
+static const char binary_alphabet[2] = "01";
+static double binary_model_root_distribution[2] = {1.0, 0.0};
+static double binary_model_mutation_matrix[4] = {0.0, 1.0, 1.0, 0.0};
+
+static mutation_model_t binary_model = {
+    .alphabet_length = 2,
+    .alphabet = binary_alphabet,
+    .root_distribution = binary_model_root_distribution,
+    .mutation_matrix = binary_model_mutation_matrix
 };
 
-static const mutation_type_t acgt_mutation_types[] = {
-    {"A", "C"},
-    {"A", "G"},
-    {"A", "T"},
-    {"C", "A"},
-    {"C", "G"},
-    {"C", "T"},
-    {"G", "A"},
-    {"G", "C"},
-    {"G", "T"},
-    {"T", "A"},
-    {"T", "C"},
-    {"T", "G"},
+static const char acgt_alphabet[4] = "ACGT";
+static double jukes_cantor_model_root_distribution[4] = {0.25, 0.25, 0.25, 0.25};
+static double jukes_cantor_model_mutation_matrix[16] =
```

Comment:
Maybe 
```c
#define ONE_THIRD (1.0 / 3.0)

{0.0, ONE_THIRD, ONE_THIRD, ONE_THIRD}
etc
``
I worry that it would be easy to end up with 10.0/3 or something in the middle of the matrix at some point and it would go unseen.